File: Sorts/MergeSort.py
"""
Brian Taylor Vann
github.com/taylor-vann

Description:
- Module to merge sort a list

Requirements:
- None

Methods:
- merge_sort
- merge
"""


def merge_sort(l):
    if len(l) <= 1:
        return l

    p = len(l) // 2

    l1 = l[:p]
    l2 = l[p:]

    l1 = merge_sort(l1)
    l2 = merge_sort(l2)

    return merge(l1, l2)


def merge(a, b):
    c = []

    while len(a) and len(b):
        if a[0] < b[0]:
            c.append(a.pop(0)) 
        else:
            c.append(b.pop(0))
            

    while len(a):
        c.append(a.pop(0))

    while len(b):
        c.append(b.pop(0))

    return c

File: ECMAScript/Sorts/MergeSort.js
/*
Brian Taylor Vann
github.com/taylor-vann

Description:
- Function to merge sort a list

Requirements:
- None

Methods:
- mergeSort
- merge
*/


function mergeSort(a) {
  if (a.length < 2) {
    return a
  }

  var p = Math.floor(a.length / 2);

  var l1 = mergeSort(a.slice(0, p));
  var l2 = mergeSort(a.slice(p));

  return merge(l1, l2);
}


function merge(a, b) {
  var n = [];

  while(a.length > 0 && b.length > 0) {
    if(a[0] < b[0]) {
      n.push(a.shift());
    }else{
      n.push(b.shift());
    }
  }

  return n.concat(a, b);
}
